### PR TITLE
Basic Metadata and Statistics

### PR DESCRIPTION
**Pull request by linbo-lin**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -2,21 +2,22 @@
...
+#![allow(clippy::borrowed_box)]
 
-mod any;
 pub mod cost;
-mod datum;
+pub mod datum;
 pub mod expression;
-mod memo;
-mod metadata;
+pub mod memo;
```
Comment (davidli2010) [inline src/lib.rs]: memo and task should not be exposed

File: src/metadata/md_provider.rs [hunk 2]
```diff
@@ -0,0 +1,5 @@
+use crate::metadata::{MdId, Metadata};
+
+pub trait MdProvider {
+    fn retrieve_metadata(&self, md_id: &Box<dyn MdId>) -> Box<dyn Metadata>;
```
Comment (davidli2010) [inline src/metadata/md_provider.rs]: Retrieving metedata may be failed, so here need to return a result

File: src/metadata/statistics.rs [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+pub type ColumnIndex = usize;
+
+/// Statistics information of a column
+#[derive(Clone, Debug)]
+pub struct ColumnStats {
+    col_idx: ColumnIndex,
+    name: String,
+    min: Datum,                   // Min value of the column
+    max: Datum,                   // Max value of the column
+    null_count: f64,              // Count of null values
```
Comment (davidli2010) [inline src/metadata/statistics.rs]: why use f64 for count, can we use u64?

File: src/metadata/statistics.rs [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+
+#[derive(Copy, Clone, Debug)]
+pub enum StorageType {
+    Heap,
+    External,
+    Virtual,
+}
+
+#[derive(Copy, Clone, Debug)]
+pub enum DistributionPolicyType {
```
Comment (davidli2010) [inline src/metadata/statistics.rs]: StorageType and DistributionPolicyType can be removed in order to simplify the business, we can consider them later.

File: src/metadata/statistics.rs (hunk 3 above)
Comment (davidli2010) [inline src/metadata/statistics.rs]: why use f64 for count, can we use u64?

File: src/metadata/statistics.rs [hunk 5]
```diff
@@ -0,0 +1,334 @@
+use crate::any::AsAny;
+use crate::datum::Datum;
+use crate::metadata::{MdId, Metadata};
+use std::any::Any;
+use std::fmt::Debug;
+use std::rc::Rc;
+
+pub trait Stats: Debug {
+    fn as_any(&self) -> &dyn Any;
```
Comment (davidli2010) [inline src/metadata/statistics.rs]: use `AsAny` trait

**Review by davidli2010 — CHANGES_REQUESTED** [review]

File: src/metadata/md_provider.rs (hunk 2 above)
Comment (linbo-lin) [reply src/metadata/md_provider.rs]: Modified to return Option

**Review by LiBinfeng-01 — APPROVED** [review]

**Review by davidli2010 — APPROVED** [review]

**Pull request merged**